### PR TITLE
Pyglet fixes for 1.4

### PR DESCRIPTION
**Pull request by oscargus**
#### References to other Issues or PRs

Closes #14748
Closes #15772 

#### Brief description of what is fixed or changed
Fixed some issues with pygletplot so that it is back working (on most/some systems).

#### Other comments
A PR aimed for the 1.4 release (a version of #15772).

#### Release Notes

* plotting
   * pygletplot is working again, although there seems to be issues on OS X.

### CODE REVIEW

**Comment by oscarbenjamin** [comment]
> Merging [#16513](https://codecov.io/gh/sympy/sympy/pull/16513?src=pr&el=desc) into [1.4](https://codecov.io/gh/sympy/sympy/commit/088ba68312c0d9154b291abdb678cb77fbd97a8d?src=pr&el=desc) will **decrease** coverage by `62.57%`.

I'm guessing this is codecov getting confused. Why does it say that coverage will go down by 60%?

**Comment by oscargus** [comment]
I was also a bit confused by that. I do not have any good explanation really...

File: sympy/utilities/runtests.py [hunk 1]
```diff
@@ -521,6 +521,9 @@ def _test(*paths, **kwargs):
     split = kwargs.get('split', None)
     time_balance = kwargs.get('time_balance', True)
     blacklist = kwargs.get('blacklist', ['sympy/integrals/rubi/rubi_tests/tests'])
+    if ON_TRAVIS:
+        # pyglet does not work on Travis
```
Comment (asmeurer) [inline sympy/utilities/runtests.py]: That's a shame.

**Comment by asmeurer** [comment]
I don't know what kind of testing you want to have here. I tried

```py
from sympy.plotting.pygletplot import PygletPlot
import sympy.plotting.pygletplot.plot_modes
from sympy.abc import x, y

PygletPlot(x*y)
```

and it worked on my Mac. It doesn't work from the normal Python shell but it does work from `isympy`. 

One issue is that when I close the plot, all keyboard input to the terminal is broken. I have to exit isympy and type `reset` to fix it. 

I ran `./bin/test pyglet` and the tests all passed. It did print several RuntimeWarnings to the terminal.

**Comment by oscarbenjamin** [comment]
This doesn't work for me on OSX. I installed pyglet using pip and
```julia
In [1]: from sympy.plotting.pygletplot import PygletPlot 
   ...: import sympy.plotting.pygletplot.plot_modes 
   ...: from sympy.abc import x, y 
   ...:  
   ...: PygletPlot(x*y)                                                                                                                        
Out[1]: [0]: x*y, 'mode=cartesian'

In [2]: 2019-04-02 21:03:06.211 Python[3189:14612259] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'nextEventMatchingMask should only be called from the Main Thread!'
*** First throw call stack:
(
	0   CoreFoundation                      0x00007fff3a03523b __exceptionPreprocess + 171
	1   libobjc.A.dylib                     0x00007fff612c5c76 objc_exception_throw + 48
	2   AppKit                              0x00007fff37ce0297 -[NSApplication(NSEvent) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 4167
	3   _ctypes.cpython-37m-darwin.so       0x000000010fcbd3cf ffi_call_unix64 + 79
	4   ???                                 0x000070001079f760 0x0 + 123145578739552
)
libc++abi.dylib: terminating with uncaught exception of type NSException
Abort trap: 6
```
It seems to do something (the screen switches to the desktop) but it crashes and closes before I see anything.

**Comment by oscargus** [comment]
I do not know either. Something that pass I guess. :-) More seriously, if I remember correctly I could not figure out how to get it to work on Travis. I do not know how/if the other plotting code is tested, but if it is possible to at least run some commands to make sure that there are no uncaught exceptions etc that may be enough.

I've got a new install on my office CentOS7 machine, so maybe the OpenGL stuff works better there now. No time to try it out yet though (although I've noticed that Matlab outputs a text claiming software OpenGL support, so I guess that drivers are lacking as the GTX1050 card I have should support it decently in hardware).

**Comment by asmeurer** [comment]
If you are using the conda Python on OS X make sure to use `pythonw`. Otherwise things that that open windows like pyglet or matplotlib cannot work properly. In my case I had to use IPython, which does some other stuff to make it work correctly.

**Comment by oscargus** [comment]
I installed this on my office computer and it worked out of the box without any error messages etc. I would probably recommend putting this into 1.4 as it works on at least some machines (and the previous version didn't work at all).

**Comment by oscarbenjamin** [comment]
Actually I've just tried on Centos and it works for me as well.

I do get an exception when trying to close the Window though:
```
Exception in thread Thread-34:
Traceback (most recent call last):
  File "/opt/rh/rh-python36/root/usr/lib64/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/opt/rh/rh-python36/root/usr/lib64/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "/home/cosc/enojb/linux/current/sympy/sympy.git/sympy/plotting/pygletplot/managed_window.py", line 73, in __event_loop__
    super(ManagedWindow, self).close()
  File "/home/cosc/enojb/linux/current/sympy/36venv/lib64/python3.6/site-packages/pyglet/window/xlib/__init__.py", line 480, in close
    self.context.destroy()
  File "/home/cosc/enojb/linux/current/sympy/36venv/lib64/python3.6/site-packages/pyglet/gl/xlib.py", line 345, in destroy
    super(XlibContext13, self).destroy()
  File "/home/cosc/enojb/linux/current/sympy/36venv/lib64/python3.6/site-packages/pyglet/gl/base.py", line 342, in destroy
    gl._shadow_window.switch_to()
  File "/home/cosc/enojb/linux/current/sympy/36venv/lib64/python3.6/site-packages/pyglet/window/xlib/__init__.py", line 499, in switch_to
    self.context.set_current()
  File "/home/cosc/enojb/linux/current/sympy/36venv/lib64/python3.6/site-packages/pyglet/gl/xlib.py", line 328, in set_current
    super(XlibContext13, self).set_current()
  File "/home/cosc/enojb/linux/current/sympy/36venv/lib64/python3.6/site-packages/pyglet/gl/base.py", line 301, in set_current
    gl_info.set_active_context()
  File "/home/cosc/enojb/linux/current/sympy/36venv/lib64/python3.6/site-packages/pyglet/gl/gl_info.py", line 98, in set_active_context
    if self.have_version(3):
  File "/home/cosc/enojb/linux/current/sympy/36venv/lib64/python3.6/site-packages/pyglet/gl/gl_info.py", line 170, in have_version
    imajor, iminor, irelease = [int(v) for v in ver.split('.', 3)[:3]]
  File "/home/cosc/enojb/linux/current/sympy/36venv/lib64/python3.6/site-packages/pyglet/gl/gl_info.py", line 170, in <listcomp>
    imajor, iminor, irelease = [int(v) for v in ver.split('.', 3)[:3]]
ValueError: invalid literal for int() with base 10: ''
```

In any case the changes here are a definite improvement. I agree that this should go into 1.4.

**Comment by oscargus** [comment]
I got that error on my earlier machine as well. It relates to the fact that the pyglet code assumes a three digit OpenGL version number, but your machine/driver only provides a two digit (so 4.5 instead of 4.5.6). A bit annoying, but nothing we can really do anything about except for catching the exception, but there's a risk that one catches other exceptions as well then...).

**Comment by oscargus** [comment]
I opened an issue at pyglet for this.

**Comment by asmeurer** [comment]
Should this be merged?

**Comment by asmeurer** [comment]
This is the last PR on the 1.4 milestone. As soon as it is merged, I will do another release candidate. Or if we decide to postpone it I can do an rc2 now.

**Comment by oscarbenjamin** [comment]
It should be merged

**Comment by oscargus** [comment]
I agree. Hence, I merge,

**Pull request merged**